chain/client/src/client.rs

```diff
@@ -1661,4 +1669,119 @@ impl Client {
         //        }
         Ok(())
     }
+
+    fn record_receive_block_timestamp(&mut self, height: BlockHeight) {
+        if let Ok(tip) = self.chain.head() {
+            self.chunks_delay_tracker.add_block_timestamp(height, tip.height);
+        }
+    }
+    fn record_receive_chunk_timestamp(&mut self, height: BlockHeight, shard_id: ShardId) {
+        if let Ok(tip) = self.chain.head() {
+            self.chunks_delay_tracker.add_chunk_timestamp(height, shard_id, tip.height);
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct HeightInfo {
+    block_received: Option<Instant>,
+    chunks_received: HashMap<ShardId, Instant>,
+}
+
+#[derive(Debug, Default)]
+struct ChunksDelayTracker {
+    heights: BTreeMap<BlockHeight, HeightInfo>,
+    prev_head_height: BlockHeight,
+}
+
+const CHUNKS_DELAY_TRACKER_HORIZON: u64 = 10;
+
+impl ChunksDelayTracker {
+    fn lowest_height(head_height: BlockHeight) -> BlockHeight {
+        head_height.saturating_sub(CHUNKS_DELAY_TRACKER_HORIZON)
+    }
+    fn remove_old_entries(&mut self, head_height: BlockHeight) {
+        let heights_to_drop: Vec<BlockHeight> = self
+            .heights
+            .iter()
```

Comment:
You want `BTreeMap::range` here: https://doc.rust-lang.org/stable/std/collections/struct.BTreeMap.html#method.range